woocommerce/woocommerce
# Remove BlockTemplateRegistry

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR removes `BlockTemplateRegistry` and associated code that is now unused as `LayoutTemplateRegistry` and the layout templates REST API endpoint are in place and being used.

In addition:

- Layout template event logging to file has been re-introduced (fixes a regression)
- Layout template event hashing has been improved to pick up more changes (fixes cases where layout template events would not be logged to file even when they had changed)
- The `woocommerce_block_template_register` action is deprecated

Closes #40620.



### How to test the changes in this Pull Request:



Put the following code in an mu-plugins file to help verify things below...

```php
<?php

use Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates\GroupInterface;

// Verify the deprecated action is still called
add_action(
	'woocommerce_block_template_register',
	function( $layout_template ) {
		error_log(
			'woocommerce_block_template_register action called for layout template ' .
			$layout_template->get_id()
		);
	}
);

// Verify the new action is called
add_action(
	'woocommerce_layout_template_after_instantiation',
	function( $layout_template_id, $layout_template_area, $layout_template ) {
		error_log(
			'woocommerce_layout_template_after_instantiation action called for layout template ' .
			$layout_template_id . '( ' . $layout_template_area . ' ) [instance correct: ' .
			( $layout_template->get_id() === $layout_template_id ? 'true' : 'false' ) . ']'
		);
	},
	10,
	3
);

// Make a layout template modification that will result in the template always
// being different from the last time it was instantiated, which will cause the
// template events to always be logged to file (with default threshold of 'warning').
add_action(
	'woocommerce_block_template_area_product-form_after_add_block_general',
	function( GroupInterface $general_group ) {
		// Ensure that a 'warning' level event is triggered that is different from the last time
		// the template was instantiated, so that the template events are different, and therefore
		// logged to file.
		throw new \Exception(
			'This exception is thrown to force a warning level event to be triggered. (timestamp: ' .
			time() .
			')'
		);
	}
);

```

With a WooCommerce dev env with the new product editor enabled (**WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**)...

1. Go to **Products** > **Add New**
2. Make sure form loads. If it does, things are good (at least as far as this PR is concerned!).
3. Check logs (`/wp-admin/admin.php?page=wc-status&tab=logs`) and make sure the code above worked correctly...

- `debug`: you should see the following three lines logged every time the product editor page is refreshed:

    - woocommerce_layout_template_after_instantiation action called for layout template simple-product( product-form ) [instance correct: true]
    - woocommerce_block_template_register is deprecated since version 8.6.0! Use woocommerce_layout_template_after_instantiation instead.
    - woocommerce_block_template_register action called for layout template simple-product

- `block_template`: you should see events logged every time the product editor page is refreshed (since the code above make sure the template is changed every time it is instantiated)



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
